Persistant storage

jedouard98 opened the pull request:
The transaction features that read in and write out transactions every 25 seconds

jedouard98 committed:
Created Transaction class

jedouard98 committed:
Added transaction class write methods

turtlefarmer commented on src/codeu/chat/server/Server.java:
```diff
@@ -46,9 +46,12 @@
   private static final Logger.Log LOG = Logger.newLog(Server.class);
 
   private static final int RELAY_REFRESH_MS = 5000;  // 5 seconds
+  private static final int TRANS_REFRESH_MS = 25000;  // 25 seconds
```
I know this is in the same format as the source code base, but I dislike using numbers for durations.  It ends up being a common source of error as variables are changed around and the value changed over time.  Consider using a Duration object or something of the sort which will make the code self-descriptive both where you define and consume the values.

turtlefarmer commented on src/codeu/chat/server/Server.java:
```diff
@@ -140,6 +150,7 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
 
         Serializers.INTEGER.write(out, NetworkCode.GET_SERVER_INFO_RESPONSE);
         Serializers.STRING.write(out, version);
+
```
Make sure you run the auto-formatter before submitting.

turtlefarmer commented on src/codeu/chat/server/Server.java:
```diff
@@ -196,11 +207,19 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
       }
     });
 
+    this.timeline.scheduleIn(TRANS_REFRESH_MS, new Runnable() {
```
this seems like the type of functionality to the Timeline class.  I would add a scheduleRecurring call to that class.

also, looking at the come of scheduleIn, the current code will schedule a new flush 25 seconds after the last flush completes, rather than every 25 seconds.  make sure this is the behavior you want, and this is probably the type of thing worth justifying in a comment.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+
+  public Transaction(Controller controller) {
+    this.controller = controller;
+    read();
```
this is an expensive blocking call.  I would recommend not having this in the constructor and requiring it to be explicitly called for flexibility, I can imagine a situation where a consumer would only want to create a new transaction without reading from existing disk.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+    read();
+  }
+
+  // takes in information that the server needs to populate itself!
```
unnecessary comment if you the function name is more descriptive.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+
+  // takes in information that the server needs to populate itself!
+  private void read() {
+    File transaction = new File("transaction.txt");
```
This class is tightly coupled to being a Chat server transaction.  I would make this more generic by having the file on disk location be a member variable, one that is populated in the constructor.  I would also have the string of "transaction.txt" be a const defined in the class that creates that Transaction object.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+import java.io.IOException;
+
+
+public final class Transaction {
```
This class actually represents a TransactionLog.  Transaction would imply a single transaction.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+  private void read() {
+    File transaction = new File("transaction.txt");
+    try {
+      Scanner in = new Scanner(transaction);
```
This works, but the Scanner class is pretty heavy for what you're doing and using a Scanner here probably isn't as efficient as using something like a BufferedReader.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+        }
+      }
+    }
+    catch (IOException e) {
```
This does not seem like the type of exception you want to swallow.  The global state of the entire server can be messed up if parsing from the transaction log fails.  I would rather the server just crash than have an inaccurate global state.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  // takes information in and stores it!
+  public void write(User user) {
```
The function itself should be more specific, such as transaction.writeCreateUser.  As you add more types of write commands in the future, the parameter type overloading will no longer be sufficient in identifying unique write functions.

Alternatively, since there's a 1:1 relationship between NetworkCode and transaction types, you could create a function which takes

write(NetCode code, Object... args)
and use a switch statement or some other mechanism inside that function to translate the write function into the right specific write function.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+
+public final class Transaction {
+
+  private ArrayList<String> transactionsList = new ArrayList<String>();
```
This is not thread safe with the current implementation.  For example, it's possible for the server to try to write to this object while it's in the flush() iterator loop.  I'm not 100% sure what the Java behavior is, but in general Collections like ArrayList will throw an exception if you try to insert while there's iterator. 

Alternatively, it's also possible for this ArrayList to insert an object while it's in the flush() block before the ArrayLlist reset operation, which will result in that command being lost forever.

This can be handled two ways.  1) locks, which is probably too heavy for what we're doing.  Or 2) Use a concurrent collection object.  Timeline.java already makes use of some concurrent friendly collections, I would look at that for examples.  For the case of transactions, since it's also FIFO, a queue should work just fine.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+
+  private void readUser(Tokenizer tokens) {
+    try {
+      tokens.next();
```
While it's nice that your commands are human readable, the way they're currently implemented add a lot of these hard-to-undesrtand token.next() calls in your code.  Here's my recommendation.

1)  In writing the commands, every single variable should be encapsulated by starting and trailing ".  This should already be well handled by the Tokenizer.
2)  Right after the readLine() call grabbing the line from the file on disk, you should process the string before handing it off to the tokenizer.  For example, if the human-readiblity helper are all in the format of "<description>:", you should strip all of these strings out.  This way, the tokenizer will only return tokens that are relevant to your command.

In addition, since the commands are so specific for each type of action right now, I would add a comment for each read command that describes the expected format of the input.

turtlefarmer commented on src/codeu/chat/server/Transaction.java:
```diff
@@ -0,0 +1,161 @@
+      Time creation = Time.fromMs(new Long(tokens.next()));
+      controller.newUser(id, name, creation);
+    }
+    catch (IOException e) {
```
Similar to what I said before, there are very few reasons exceptions should ever be swallowed.  I don't think this is a good reason to swallow exceptions here or in any of the read commands.

jedouard98 committed:
Updated Transaction to Transaction Log

-fixed name to be more clear
-Utilizes LinkedList Queue
-Utilizes BufferedReader
-Added TODO comments for late version fixes

jedouard98 committed:
Updated Server according to pull comments

-Renamed transaction write functions
-Added the newly updated Timeline function

jedouard98 committed:
Updated Server with correct code from last commit

jedouard98 committed:
Updated Timeline with Reccurring function
